src/app/clusters/commodity-metering-server/commodity-metering-server.cpp

```diff
@@ -0,0 +1,360 @@
+
+#include <protocols/interaction_model/StatusCode.h>
+
+#include <app/AttributeAccessInterface.h>
+#include <app/AttributeAccessInterfaceRegistry.h>
+#include <app/EventLogging.h>
+#include <app/reporting/reporting.h>
+#include <app/util/attribute-storage.h>
+#include <lib/support/CodeUtils.h>
+
+using namespace chip;
+using namespace chip::app;
+using namespace chip::app::DataModel;
+using namespace chip::app::Clusters;
+using namespace chip::app::Clusters::Globals;
+using namespace chip::app::Clusters::Globals::Structs;
+using namespace chip::app::Clusters::CommodityMetering;
+using namespace chip::app::Clusters::CommodityMetering::Attributes;
+
+using chip::Protocols::InteractionModel::Status;
+
+namespace chip {
+namespace app {
+namespace Clusters {
+namespace CommodityMetering {
+
+// Some constraints for lists limitation
+constexpr uint8_t kMaxMeteredQuantityEntries                    = 128;
+constexpr uint8_t kMaxTariffComponentIDsPerMeteredQuantityEntry = 128;
+
+namespace {
+
+inline bool operator==(const Span<const uint32_t> & a, const Span<const uint32_t> & b)
+{
+    return std::equal(a.begin(), a.end(), b.begin());
+}
+
+inline bool operator==(const Structs::MeteredQuantityStruct::Type & lhs, const Structs::MeteredQuantityStruct::Type & rhs)
+{
+    return (lhs.tariffComponentIDs == rhs.tariffComponentIDs) && (lhs.quantity == rhs.quantity);
+}
+
+template <typename T>
+bool NullableListEqual(const DataModel::Nullable<DataModel::List<T>> & a, const DataModel::Nullable<DataModel::List<T>> & b)
+{
+    if (a.IsNull() || b.IsNull())
+    {
+        return a.IsNull() == b.IsNull();
+    }
+    else if (a.Value().size() == b.Value().size())
+    {
+        for (size_t i = 0; i < a.Value().size(); i++)
+        {
+            if (a.Value()[i] == b.Value()[i])
+            {
+                continue;
+            }
+            return false;
+        }
+        return true;
+    }
+    return false;
+}
+
+template <typename T>
+struct SpanCopier
+{
+    static bool Copy(const chip::Span<const T> & source, DataModel::List<const T> & destination,
+                     size_t maxElements = std::numeric_limits<size_t>::max())
+    {
+        if (source.empty())
+        {
+            destination = DataModel::List<const T>();
+            return true;
+        }
+
+        size_t elementsToCopy = std::min(source.size(), maxElements);
+        auto * buffer         = static_cast<T *>(chip::Platform::MemoryCalloc(elementsToCopy, sizeof(T)));
+
+        if (!buffer)
+        {
+            return false;
+        }
+
+        std::copy(source.begin(), source.begin() + elementsToCopy, buffer);
+        destination = DataModel::List<const T>(chip::Span<const T>(buffer, elementsToCopy));
+        return true;
+    }
+};
+} // namespace
+
+static void CleanUpIDs(DataModel::List<const uint32_t> & IDs)
+{
+    if (!IDs.empty() && IDs.data())
+    {
+        Platform::MemoryFree(const_cast<uint32_t *>(IDs.data()));
```

Comment:
Please document why this const_cast is OK (i.e. because we know we allocated the buffer ourselves).